lib/phoenix/endpoint/cowboy2_handler.ex

```diff
@@ -113,9 +102,43 @@ defmodule Phoenix.Endpoint.Cowboy2Handler do
     {id, start, restart, shutdown, type, modules}
   end
 
-  defp default_for(Phoenix.Transports.LongPoll), do: Plug.Adapters.Cowboy2.Handler
-  defp default_for(Phoenix.Transports.WebSocket), do: Phoenix.Endpoint.Cowboy2WebSocket
-  defp default_for(_), do: nil
+  @doc false
+  def build_dispatches(endpoint, cowboy_handler, plug_handler) do
+    socket_dispatches =
+      for {path, socket, opts} <- endpoint.__sockets__,
+          {transport, {module, config}} <- socket.__transports__,
+          # Allow handlers to be configured at the transport level
+          handler = config[:cowboy] || default_for(cowboy_handler, module),
```

Comment:
Could we pass a map into this function instead to avoid referencing the `Phoenix.Endpoint.CowboyHander` in `default_for`?

something like:

```elixir
handler = config[:cowboy] || Map.get(default_handlers, module),
```


```elixir
  default_handlers = %{
    Phoenix.Transports.LongPoll => Plug.Adapters.Cowboy2.Handler,
   Phoenix.Transport.WebSocket => Phoenix.Endpoint.Cowboy2WebSocket
  }
dispatches = build_dispatches(endpoint, __MODULE__, Plug.Adapters.Cowboy2.Handler, default_handlers)
```

Then in the Cowboy1 handler we can have:

```elixir
  default_handlers = %{
    Phoenix.Transports.LongPoll => Plug.Adapters.Cowboy.Handler,
   Phoenix.Transport.WebSocket => Phoenix.Endpoint.CowboyWebSocket
  }
  Cowboy2Handler.build_dispatches(endpoint, __MODULE__, Plug.Adapters.Cowboy.Handler, default_handlers)
```

Reply:
Good call. Just remember to `@doc false` it. It would help us write some unit tests too!